### PR TITLE
Add example and tutorial to explain how to create custom metrics for Prometheus

### LINKED ISSUE #907: [Feature] Add an example and tutorial to explain how to create custom metrics for Prometheus
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/kuberay/issues) and found no similar feature requirement.

### Description

Add an example and tutorial to explain how to create custom metrics for Prometheus.

https://github.com/ray-project/kuberay/blob/master/docs/guidance/prometheus-grafana.md#custom-metrics--alerting

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by Yicheng-Lu-llll**
## Why are these changes needed?
Add an example and tutorial to explain how to create custom metrics for Prometheus.

## Related issue number
Closes #907

## Checks

Using Step 9 to improve:
Run` ./install/prometheus/install.sh `without installing KubeRay operator and RayCluster:
In 5 minutes:
f4d8d9a8b9f8c37bd5f70979901398c
After 5 minutes:
67d282e3ea76009429773c840fbf05d

After installing KubeRay operator and RayCluster:
cb6d8db4ff224302b7ae6d01537a044

79592c42616d21363cfebca6bc0d379

### CODE REVIEW

File: docs/guidance/prometheus-grafana.md [hunk 1]
```diff
@@ -160,19 +163,124 @@ spec:
...
+      expr: |
+      (
+        100 * (
+          sum(rate(ray_gcs_update_resource_usage_time_bucket{container="ray-head", le="20.0"}[30d]))
+          /
+          sum(rate(ray_gcs_update_resource_usage_time_count{container="ray-head"}[30d]))
+        )
+      )
+```
+* The recording rule above is one of rules defined in [prometheusRules.yaml](../../config/prometheus/rules/prometheusRules.yaml), and it is created by **install.sh**, Hence, no need to create anything here.
```
Comment (kevin85421) [inline docs/guidance/prometheus-grafana.md:204]:
```suggestion
* The recording rule above is one of rules defined in [prometheusRules.yaml](../../config/prometheus/rules/prometheusRules.yaml), and it is created by **install.sh**. Hence, no need to create anything here.
```

File: docs/guidance/prometheus-grafana.md [hunk 2]
```diff
@@ -160,19 +163,124 @@ spec:
...
+
+* See [PrometheusRule official document](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#prometheusrule) for more details about the configurations.
+
+* `release: $HELM_RELEASE`: Prometheus can only detect PrometheusRule with this label.
```
Comment (kevin85421) [inline docs/guidance/prometheus-grafana.md:208]: Point to the above section.

File: docs/guidance/prometheus-grafana.md [hunk 3]
```diff
@@ -160,19 +163,124 @@ spec:
...
+  #   matchLabels:
+  #     release: prometheus
+  ```
+
+* PrometheusRule can be reloaded at runtime. Use `kubectl apply {modified prometheusRules.yaml}` to reconfigure the rules if needed.
+
+
+## Step 8: Define Alert Conditions with Alerting Rules
+
+[Alerting rules](https://prometheus.io/docs/prometheus/latest/configuration/alerting_rules/) allow us to define alert conditions based on [PromQL](https://prometheus.io/docs/prometheus/latest/querying/basics/) expressions and to send notifications about firing alerts to [Alertmanager](https://prometheus.io/docs/alerting/latest/alertmanager).
```
Comment (kevin85421) [inline docs/guidance/prometheus-grafana.md:221]: Explain more about AlertManager.

File: docs/guidance/prometheus-grafana.md [hunk 4]
```diff
@@ -160,19 +163,124 @@ spec:
...
+  labels:
+    # `release: $HELM_RELEASE`: Prometheus can only detect Rule with this label.
+    release: prometheus
+spec:
+  groups:
+  - name: ray-cluster-main-staging-gcs.rules
+    # How often rules in the group are evaluated.
+    interval: 30s
+    rules:
+      # The name of the alert.
```
Comment (kevin85421) [inline docs/guidance/prometheus-grafana.md:238]: Remove.

File: docs/guidance/prometheus-grafana.md [hunk 5]
```diff
@@ -160,19 +163,124 @@ spec:
...
+    - alert: RayGlobalControlStoreShortWindowSLO
+      # A set of informational labels that can be used to store longer additional information.
+      annotations:
+        description: Ray GCS Resource Update requests are above the SLO (99.9 > 20ms).
+          Burning rate is 14.4 for short window (5m & 1h)
+        summary: Ray GCS is burning too much error budget (short window)
+      # The PromQL expression to evaluate.
```
Comment (kevin85421) [inline docs/guidance/prometheus-grafana.md:245]:
```suggestion
      # PromQL expression
```

File: docs/guidance/prometheus-grafana.md [hunk 6]
```diff
@@ -160,19 +163,124 @@ spec:
...
+                        ray_gcs_error_budget_5m > 14.4*0.1
+                      )
+                      and
+                      (
+                        ray_gcs_error_budget_1h > 14.4*0.1
+                      )
+      # time to wait between first encountering a new expression output vector element and counting an alert as firing for this element.
+      for: 5m
+      # A set of additional labels to be attached to the alert.
+      labels:
```
Comment (kevin85421) [inline docs/guidance/prometheus-grafana.md:257]: What's the difference between `labels` and `annotations`?

File: docs/guidance/prometheus-grafana.md [hunk 7]
```diff
@@ -160,19 +163,124 @@ spec:
...
+      # The PromQL expression to evaluate.
+      expr:
+                      (
+                        ray_gcs_error_budget_5m > 14.4*0.1
+                      )
+                      and
+                      (
+                        ray_gcs_error_budget_1h > 14.4*0.1
+                      )
+      # time to wait between first encountering a new expression output vector element and counting an alert as firing for this element.
```
Comment (kevin85421) [inline docs/guidance/prometheus-grafana.md:254]: Hard to understand.

File: docs/guidance/prometheus-grafana.md [hunk 8]
```diff
@@ -160,19 +163,124 @@ spec:
...
 ```
 
+- Go to `${YOUR_IP}:9090/targets` (e.g. `127.0.0.1:9090/targets`). You should be able to see:
+  - `podMonitor/prometheus-system/ray-workers-monitor/0 (1/1 up)`
+  - `serviceMonitor/prometheus-system/ray-head-monitor/0 (1/1 up)`
 ![Prometheus Web UI](../images/prometheus_web_ui.png)
 
-## Step 8: Access Grafana
+- Go to `${YOUR_IP}:9090/targets`. You should be able to query:
+  - [System Metrics](https://docs.ray.io/en/latest/ray-observability/ray-metrics.html#system-metrics )
```
Comment (kevin85421) [inline docs/guidance/prometheus-grafana.md:276]:
```suggestion
  - [System Metrics](https://docs.ray.io/en/latest/ray-observability/ray-metrics.html#system-metrics)
```

File: docs/guidance/prometheus-grafana.md [hunk 9]
```diff
@@ -160,19 +163,124 @@ spec:
...
+kind: PrometheusRule
+metadata:
+  name: ray-cluster-gcs-rules
+  namespace: prometheus-system
+  labels:
+    # `release: $HELM_RELEASE`: Prometheus can only detect Rule with this label.
+    release: prometheus
+spec:
+  groups:
+  - # Name of the group. Rules within a group are run sequentially at a regular interval,
```
[the remark this answers is not in the document]
Comment (Yicheng-Lu-llll) [reply docs/guidance/prometheus-grafana.md]: They are actually referring to the same thing. Rules within a group are run sequentially at the same evaluation interval. regular interval and same evaluation time all refer to the same evaluation interval.

File: docs/guidance/prometheus-grafana.md [hunk 10]
```diff
@@ -95,19 +95,21 @@ spec:
 * The YAML example above is [serviceMonitor.yaml](../../config/prometheus/serviceMonitor.yaml), and it is created by **install.sh**. Hence, no need to create anything here.
 * See [ServiceMonitor official document](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#servicemonitor) for more details about the configurations.
 
-* `release: $HELM_RELEASE`: Prometheus can only detect ServiceMonitor with this label.
+* #### `release: $HELM_RELEASE`: Prometheus can only detect ServiceMonitor with this label.
```
Comment (kevin85421) [inline docs/guidance/prometheus-grafana.md:98]: What's the purpose of "####"?

File: docs/guidance/prometheus-grafana.md [hunk 11]
```diff
@@ -151,6 +153,8 @@ spec:
   podMetricsEndpoints:
   - port: metrics
 ```
+* `release: $HELM_RELEASE`: Prometheus can only detect PodMonitor with this label. See [here](#release-helm_release-prometheus-can-only-detect-servicemonitor-with-this-label ) for more details.
```
Comment (kevin85421) [inline docs/guidance/prometheus-grafana.md:156]:
```suggestion
* `release: $HELM_RELEASE`: Prometheus can only detect PodMonitor with this label. See [here](#release-helm_release-prometheus-can-only-detect-servicemonitor-with-this-label) for more details.
```

File: docs/guidance/prometheus-grafana.md [hunk 12]
```diff
@@ -160,19 +164,119 @@ spec:
 
 * `ray.io/cluster: $RAY_CLUSTER_NAME`: We also define `metadata.labels` by manually adding `ray.io/cluster: <ray-cluster-name>` and then instructing the PodMonitors resource to add that label in the scraped metrics via `spec.podTargetLabels[0].ray.io/cluster`.
 
-## Step 7: Access Prometheus Web UI
+## Step 7: Collect Custom metrics with Recording Rules
```
Comment (kevin85421) [inline docs/guidance/prometheus-grafana.md:167]:
```suggestion
## Step 7: Collect custom metrics with Recording Rules
```

File: docs/guidance/prometheus-grafana.md [hunk 13]
```diff
@@ -160,19 +167,119 @@ spec:
...
+kind: PrometheusRule
+metadata:
+  name: ray-cluster-gcs-rules
+  namespace: prometheus-system
+  labels:
+    # `release: $HELM_RELEASE`: Prometheus can only detect Recording Rules with this label.
+    release: prometheus
+spec:
+  groups:
+- # Rules within a group are run sequentially at the same evaluation interval.
```
Comment (kevin85421) [inline docs/guidance/prometheus-grafana.md:185]:
```suggestion
- # Rules within a group will be triggered periodically with the same evaluation interval (30s in this example).
```

File: docs/guidance/prometheus-grafana.md [hunk 14]
```diff
@@ -160,19 +167,119 @@ spec:
...
+        description: Ray GCS is not emitting any metrics for Resource Update requests
+        summary: Ray GCS is not emitting metrics anymore
+      # PromQL expression.
+      expr: |
+                      (
+                       absent(ray_gcs_update_resource_usage_time_bucket) == 1
+                      )
+      # Time that Prometheus will wait and check if the alert continues to be active during each evaluation before firing the alert.
+      # firing alerts may be due to false positives or noise if the setting value is too small.
+      # firing alerts may not be handled in time if the setting value is too big.
```
Comment (kevin85421) [inline docs/guidance/prometheus-grafana.md:249]:
```suggestion
      # On the other hand, if the value is big, the alerts may not be handled in time.
```

**Review by kevin85421 — APPROVED** [review]
LGTM

**Pull request merged**